src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerDriver.cs

```diff
@@ -1107,9 +1114,48 @@ private bool ShouldExecuteCodeBlockActions(AnalysisScope analysisScope, ISymbol
                 return computeDeclarationAnalysisData();
             }
 
-            return _declarationAnalysisDataCache.GetValue(declaration, decl => computeDeclarationAnalysisData());
+            var map = _declarationAnalysisDataCache.GetValue(compilation, _ => new Dictionary<Location, DeclarationAnalysisData>());
+            var location = declaration.GetLocation();
+
+            DeclarationAnalysisData data;
+            lock (_gate)
+            {
+                if (map.TryGetValue(location, out data))
+                {
+                    return data;
+                }
+            }
+
+            data = computeDeclarationAnalysisData();
+
+            lock (_gate)
```

Comment:
A statement of that guarantee should appear in a comment somewhere in this code.